pkg/dyninst/integration_test.go

```diff
@@ -0,0 +1,198 @@
+	require.NoError(t, err)
+	require.NotNil(t, codeDump)
+	defer func() { require.NoError(t, codeDump.Close()) }()
+
+	bpfObj, err := compiler.CompileBPFProgram(*irp, codeDump)
+	require.NoError(t, err)
+	require.NotNil(t, bpfObj)
+	defer func() { bpfObj.Close() }()
+
+	bpfObjDump, err := os.Create(filepath.Join(tempDir, "probe.bpf.o"))
+	require.NoError(t, err)
+	require.NotNil(t, bpfObjDump)
+	defer func() { require.NoError(t, bpfObjDump.Close()) }()
+	_, err = io.Copy(bpfObjDump, bpfObj)
+	require.NoError(t, err)
+
+	spec, err := ebpf.LoadCollectionSpecFromReader(bpfObj)
+	require.NoError(t, err)
+	require.NotNil(t, spec)
+
+	bpfCollection, err := ebpf.NewCollectionWithOptions(spec, ebpf.CollectionOptions{})
+	require.NoError(t, err)
+	require.NotNil(t, bpfCollection)
+	defer func() { bpfCollection.Close() }()
+
+	bpfProg, ok := bpfCollection.Programs["probe_run_with_cookie"]
+	require.True(t, ok)
+	require.NotNil(t, bpfProg)
+
+	sampleLink, err := link.OpenExecutable(sampleServicePath)
+	require.NoError(t, err)
+	require.NotNil(t, sampleLink)
+
+	// Launch the sample service, inject the BPF program and collect the output.
+	t.Logf("running and instrumenting sample")
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	sampleProc := exec.CommandContext(ctx, sampleServicePath)
+	require.NotNil(t, sampleProc)
+	sampleStdin, err := sampleProc.StdinPipe()
+	require.NoError(t, err)
+	sampleProc.Stdout, err = os.Create(filepath.Join(tempDir, "sample.out"))
+	require.NoError(t, err)
+	sampleProc.Stderr, err = os.Create(filepath.Join(tempDir, "sample.err"))
+	require.NoError(t, err)
+	require.NotNil(t, sampleStdin)
+	err = sampleProc.Start()
+	require.NoError(t, err)
+
+	bpfProbe, err := sampleLink.Uprobe(
+		"main.intArg",
+		bpfProg,
+		&link.UprobeOptions{
+			PID: os.Getpid(),
+		},
+	)
+	require.NoError(t, err)
+	require.NotNil(t, bpfProbe)
+	defer func() { require.NoError(t, bpfProbe.Close()) }()
+
+	attached, err := sampleLink.Uprobe("main.intArg", bpfProg, &link.UprobeOptions{
+		PID:    sampleProc.Process.Pid,
+		Cookie: 0,
+	})
+	require.NoError(t, err)
+	require.NotNil(t, attached)
+	defer func() { require.NoError(t, attached.Close()) }()
+
+	// Trigger the function calls.
+	sampleStdin.Write([]byte("\n"))
+
+	err = sampleProc.Wait()
+	require.NoError(t, err)
+
+	// Validate the output. For now we just check the total length.
+	t.Logf("processing output")
+	rd, err := ringbuf.NewReader(bpfCollection.Maps["out_ringbuf"])
+	require.NoError(t, err)
+	require.NotNil(t, rd)
+
+	bpfOutDump, err := os.Create(filepath.Join(tempDir, "probe.bpf.out"))
+	require.NoError(t, err)
+	require.NotNil(t, bpfOutDump)
+	defer func() { require.NoError(t, bpfOutDump.Close()) }()
+
+	require.Greater(t, rd.AvailableBytes(), 0)
+	record, err := rd.Read()
+	require.NoError(t, err)
+	require.NotNil(t, record)
+	bpfOutDump.Write(record.RawSample)
+
+	require.Equal(t, 88, len(record.RawSample))
```

Comment:
nit: write out the math on this one? `unsafe.Sizeof(output.EventHeader)+unsafe.Sizeof(output.DataItemHeader)+...`?